tests/Benchmarks/BufferWriterBench.cs

```diff
@@ -0,0 +1,527 @@
+    {
+        private T _output;
+        private Span<byte> _span;
+        private int _buffered;
+
+        public BufferWriter(T output)
+        {
+            _buffered = 0;
+            _output = output;
+            _span = output.GetSpan();
+        }
+
+        public Span<byte> Span => _span;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Commit()
+        {
+            var buffered = _buffered;
+            if (buffered > 0)
+            {
+                _buffered = 0;
+                _output.Advance(buffered);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Advance(int count)
+        {
+            _buffered += count;
+            _span = _span.Slice(count);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Write(ReadOnlySpan<byte> source)
+        {
+            if (_span.Length >= source.Length)
+            {
+                source.CopyTo(_span);
+                Advance(source.Length);
+            }
+            else
+            {
+                WriteMultiBuffer(source);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Ensure(int count = 1)
+        {
+            if (_span.Length < count)
+            {
+                EnsureMore(count);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void EnsureMore(int count = 0)
+        {
+            if (_buffered > 0)
+            {
+                Commit();
+            }
+
+            _output.GetMemory(count);
+            _span = _output.GetSpan();
+        }
+
+        private void WriteMultiBuffer(ReadOnlySpan<byte> source)
+        {
+            while (source.Length > 0)
+            {
+                if (_span.Length == 0)
+                {
+                    EnsureMore();
+                }
+
+                var writable = Math.Min(source.Length, _span.Length);
+                source.Slice(0, writable).CopyTo(_span);
+                source = source.Slice(writable);
+                Advance(writable);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Write(ulong number)
+        {
+            // Try to format directly
+            if (Utf8Formatter.TryFormat(number, Span, out int bytesWritten))
+            {
+                Advance(bytesWritten);
+            }
+            else
+            {
+                // Ask for at least 20 bytes
+                Ensure(20);
+
+                Debug.Assert(Span.Length >= 20, "Buffer is < 20 bytes");
+
+                // Try again
+                if (Utf8Formatter.TryFormat(number, Span, out bytesWritten))
+                {
+                    Advance(bytesWritten);
+                }
+            }
+        }
+    }
+}
+
+// copy from System.Buffers.ReaderWriter to isolate cross-dll calls.
```

Comment:
Why is this necessary?

Reply:
For some reason source code compiled into the test is a bit faster. I wanted to have totally apples to apples comparison between BufferWriters and so I wanted to have them both in the same project.